src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+//SPDX-License-Identifier: None
+pragma solidity ^0.8.0;
+
+import "src/interfaces/IMarket.sol";
+import "src/interfaces/IERC20.sol";
+import "src/interfaces/IDola.sol";
+import "src/interfaces/ITransformHelper.sol";
+import {Ownable} from "lib/openzeppelin-contracts/contracts/access/Ownable.sol";
+import {ReentrancyGuard} from "lib/openzeppelin-contracts/contracts/security/ReentrancyGuard.sol";
+
+// Accelerated leverage engine
+contract ALE is Ownable, ReentrancyGuard {
+    error CollateralNotSet();
+    error MarketNotSetForCollateral(address collateral);
+    error SwapFailed();
+    error DOLAInvalidBorrow();
+    error DOLAInvalidRepay();
+    error InvalidProxyAddress();
+
+    // 0x ExchangeProxy address.
+    // See https://docs.0x.org/developer-resources/contract-addresses
+    address payable public exchangeProxy;
+
+    IDola public immutable dola;
+
+    struct Market {
+        address market;
+        address helper;
+        address collateral;
+    }
+
+    struct Permit {
+        uint256 deadline;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    // Mapping of sellToken/buyToken to Market structs
+    // NOTE: in normal cases sellToken/buyToken is the collateral token,
+    // in other cases it could be different (eg. st-yCRV is collateral, yCRV is the token to be swapped from/to DOLA)
+    // That's why we need to keep track of collateral in market struct
+    mapping(address => Market) public markets;
+
+    constructor(address _dola, address _exchangeProxy) Ownable(msg.sender) {
+        dola = IDola(_dola);
+        exchangeProxy = payable(address(_exchangeProxy));
+    }
+
+    function setExchangeProxy(address _exchangeProxy) external onlyOwner {
+        if (_exchangeProxy == address(0)) revert InvalidProxyAddress();
+        exchangeProxy = payable(_exchangeProxy);
+    }
+
+    function setMarket(
+        address _market,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].market = _market;
+    }
+
+    function setMarketHelper(
+        address _helper,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].helper = _helper;
+    }
+
+    function setMarketCollateral(
+        address _collateral,
+        address _buySelltoken
+    ) external onlyOwner {
+        markets[_buySelltoken].collateral = _collateral;
+    }
+
+    function leveragePosition(
+        // Amount of DOLA to borrow
+        uint256 _value,
+        // `buyTokenAddress` field from the API response.
+        address _buyTokenAddress,
+        // The `allowanceTarget` field from the API response.
+        address _spender,
+        // The `data` field from the API response.
+        bytes calldata _swapCallData,
+        // Permit data
+        Permit calldata _permit,
+        // Optional helper data in case the collateral needs to be transformed
+        bytes calldata _helperData
```

Comment:
Reformat comments to natSpec doc

Reply:
Fixed in [21265a2](https://github.com/InverseFinance/FiRM/pull/52/commits/21265a282ef6b19fa53c46e5b48fe83f48880686)